File: Android/app/src/main/java/com/dreamproject/inwords/core/util/Event.java
package com.dreamproject.inwords.core.util;

/**
 * Used as a wrapper for data that is exposed via a LiveData that represents an event.
 */
public class Event<T> {
    private T content;
    private boolean hasBeenHandled = false;

    public Event(T content) {
        this.content = content;
    }

    /**
     * Returns the content and prevents its use again.
     */
    public T getContentIfNotHandled() {
        if (hasBeenHandled) {
            return null;
        } else {
            hasBeenHandled = true;
            return content;
        }
    }

    public boolean handle() {
        if (hasBeenHandled) {
            return false;
        } else {
            return hasBeenHandled = true;
        }
    }

    /**
     * Returns the content, even if it's already been handled.
     */
    public T peekContent() {
        return content;
    }
}

File: Android/app/src/main/java/ru/inwords/inwords/core/util/Event.kt
package ru.inwords.inwords.core.util

/**
 * Used as a wrapper for data that is exposed via a LiveData that represents an event.
 */
class Event<T : Any>(private val content: T) {
    private var hasBeenHandled = false

    /**
     * Returns the content and prevents its use again.
     */
    val contentIfNotHandled: T?
        get() {
            return if (hasBeenHandled) {
                null
            } else {
                hasBeenHandled = true
                content
            }
        }

    fun handle(): Boolean {
        return if (hasBeenHandled) {
            false
        } else {
            hasBeenHandled = true
            true
        }
    }

    /**
     * Returns the content, even if it's already been handled.
     */
    fun peekContent(): T {
        return content
    }
}
